### ISSUE TITLE
Support for cdef class templating

### ISSUE THREAD

**Issue by leftys** [cython/cython#3283]
Hello, I am implementing a efficient numeric circular buffer 'library' in Cython and I was surprised that it's not possible to use templating with extension types (cdef class), only methods can be templated. This way, it's impossible to implement any generic classes or containers! Is that true? If so, it would be a very useful new feature. Also if you know about a nice workaround, I would be glad to hear about it.

**Comment by McSinyx** [comment]
Could you please provide an example of what you're trying to achieve?

**Comment by leftys** [comment]
```python
cdef class CircularBufferInt:
    ''' 
    Efficient cythonized implementation of circular buffer for numeric types.

    Cache friendly implementation. Support for fixed size or automatic expansion and memory reallocation.
    '''
    cdef:
        np.int[:] _buffer 
        long _last_position
        long _first_position
    cdef readonly:
        long size
        long pushed_count
        bint autoexpand
    
    def __init__(self, size = 10, *, autoexpand = True):
        if size >= 2**32:
            raise ValueError(size)
        self._buffer = np.zeros(size, dtype = np.int)
        self._last_position = 0
        self._first_position = 0
        self.pushed_count = 0
        self.size = size
        self.autoexpand = autoexpand
        
    cpdef void push_back(self, np.int value):
        if self.autoexpand and self.pushed_count == self.size:
            self.expand(2 * self.size)

        self._buffer[self._last_position] = value
        self._last_position += 1
        self._last_position %= self.size
        if self._last_position == self._first_position:
            self._first_position += 1
        self._first_position %= self.size
        self.pushed_count += 1
        if self.pushed_count > self.size:
            self.pushed_count = self.size
    ...
```
I would like to use any numeric format instead of the `np.int` in `_buffer` and `push_back` `value`.

**Comment by scoder** [comment]
Extension types that use fused types in their attributes are not easy to support since they (may) lead to different object struct layouts for the same type. Or, they would require some form of internal subtype management, which is also not easy to support and almost certainly user visible.

Consider asking on the cython-users mailing list how others have solved this use case.

**Comment by leftys** [comment]
I though this is exactly the problem C++ templates are solving and it would be sufficient to just use them in the resulting C++ code. Imho templates are one of the strongest features of C++ as they allow writing very fast code while keeping generality and nice design without duplicating code. 

Btw, iirc templates are already supported in pxd files when importing templated C++ classes, but that's probably an entirely different problem.

**Comment by scoder** [comment]
You can use C++ templates, sure. But you can't use memory views (or buffers), because they require code generation support by Cython.

I'll close this ticket as a duplicate of #2120, please ask on the mailing list if you need further help.

**Comment by scoder** [comment]
Actually not a duplicate of what I though it was, but I can't find a ticket for this feature request, so let's keep this open.

**Comment by scoder** [comment]
See https://github.com/cython/cython/issues/3283#issuecomment-568553780 above.

**Comment by adam2392** [comment]
@leftys were you ever able to solve this problem on the Cython mailing list? If so, would you be able to link me?

I also have an issue where cython fused types for class attributes are not supported. This forces me to essentially make a copy paste of an entire class along with its methods just because one class attribute has a different type.

**Comment by jjerphan** [comment]
[Tempita](https://github.com/cython/cython/blob/master/Cython/Tempita/_tempita.py) might be useful. This is what is used within scikit-learn to define several extension types for several dtypes; as an example see https://github.com/scikit-learn/scikit-learn/pull/20481.

**Comment by scoder** [comment]
Also see https://github.com/cython/cython/issues/2596#issuecomment-1059048871
Feedback and PR welcome.

**Comment by erezinman** [comment]
@scoder For syntax I suggest something along the lines of generics. That is:

```cython
cdef class MyFusedClass[fused_type](MyBaseClass): ...
```

This is similar to the syntax changes in Python 3.12 ([PEP695](https://peps.python.org/pep-0695/))

In Pure Python mode it can be something like:

```python
class MyFusedClass(cython.FusedClass[fused_type], MyBasedClass):
    ...
```

The implementation in Pure Python is rather simple and goes like:
```python
from types import new_class

class FusedClass:
    __args__ = None

    def __new__(cls, *args, **kwargs):
        if cls.__args__ is None:
            raise Exception(...)
        return super().__new__(cls)
    
    def __class_getitem__(cls, args):
        # TODO: Validate args
        if not isinstance(args, tuple):
            args = (args, )
        derived = new_class(f'{cls.__name__}[{", ".join(map(repr, args))}]', bases=(cls, ))
        derived.__args__ = args
        
        return derived

FusedClass[1, 2, 'adasfd']()
#  <types.FusedClass[1, 2, 'adasfd'] object at 0x7f35ff96ab50>

```

**Comment by scoder** [comment]
I don't think we need new syntax for the declaration side. That part is clear from the mere fact that fused types are being used as attributes.

But yes, for the object creation side, generics seem reasonable.
